{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clustering simulated sequences using TMRCA\n",
    "\n",
    "Name: CQS21 \n",
    "FYP 2024"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting msprime\n",
      "  Using cached msprime-1.3.1.tar.gz (888 kB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Requirement already satisfied: numpy in c:\\python310\\lib\\site-packages (from msprime) (1.26.4)\n",
      "Collecting newick>=1.3.0 (from msprime)\n",
      "  Using cached newick-1.9.0-py2.py3-none-any.whl.metadata (8.4 kB)\n",
      "Requirement already satisfied: tskit>=0.5.2 in c:\\python310\\lib\\site-packages (from msprime) (0.5.6)\n",
      "Collecting demes>=0.2 (from msprime)\n",
      "  Using cached demes-0.2.3-py3-none-any.whl.metadata (1.9 kB)\n",
      "Requirement already satisfied: attrs>=20.3.0 in c:\\python310\\lib\\site-packages (from demes>=0.2->msprime) (23.2.0)\n",
      "Collecting ruamel.yaml>=0.15.78 (from demes>=0.2->msprime)\n",
      "  Using cached ruamel.yaml-0.18.6-py3-none-any.whl.metadata (23 kB)\n",
      "Requirement already satisfied: jsonschema>=3.0.0 in c:\\python310\\lib\\site-packages (from tskit>=0.5.2->msprime) (4.21.1)\n",
      "Requirement already satisfied: svgwrite>=1.1.10 in c:\\python310\\lib\\site-packages (from tskit>=0.5.2->msprime) (1.4.3)\n",
      "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in c:\\python310\\lib\\site-packages (from jsonschema>=3.0.0->tskit>=0.5.2->msprime) (2023.12.1)\n",
      "Requirement already satisfied: referencing>=0.28.4 in c:\\python310\\lib\\site-packages (from jsonschema>=3.0.0->tskit>=0.5.2->msprime) (0.34.0)\n",
      "Requirement already satisfied: rpds-py>=0.7.1 in c:\\python310\\lib\\site-packages (from jsonschema>=3.0.0->tskit>=0.5.2->msprime) (0.18.0)\n",
      "Collecting ruamel.yaml.clib>=0.2.7 (from ruamel.yaml>=0.15.78->demes>=0.2->msprime)\n",
      "  Using cached ruamel.yaml.clib-0.2.8-cp310-cp310-win_amd64.whl.metadata (2.3 kB)\n",
      "Using cached demes-0.2.3-py3-none-any.whl (40 kB)\n",
      "Using cached newick-1.9.0-py2.py3-none-any.whl (15 kB)\n",
      "Using cached ruamel.yaml-0.18.6-py3-none-any.whl (117 kB)\n",
      "Using cached ruamel.yaml.clib-0.2.8-cp310-cp310-win_amd64.whl (117 kB)\n",
      "Building wheels for collected packages: msprime\n",
      "  Building wheel for msprime (pyproject.toml): started\n",
      "  Building wheel for msprime (pyproject.toml): finished with status 'error'\n",
      "Failed to build msprime\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: Ignoring invalid distribution -p (c:\\python310\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution - (c:\\python310\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -p (c:\\python310\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -0p (c:\\python310\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -1p (c:\\python310\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -2p (c:\\python310\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -3p (c:\\python310\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -4p (c:\\python310\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -5p (c:\\python310\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -6p (c:\\python310\\lib\\site-packages)\n",
      "WARNING: Ignoring invalid distribution -ip (c:\\python310\\lib\\site-packages)\n",
      "  error: subprocess-exited-with-error\n",
      "  \n",
      "  × Building wheel for msprime (pyproject.toml) did not run successfully.\n",
      "  │ exit code: 1\n",
      "  ╰─> [73 lines of output]\n",
      "      <string>:27: UserWarning: Error occured getting GSL path config: [WinError 2] The system cannot find the file specified\n",
      "      running bdist_wheel\n",
      "      running build\n",
      "      running build_py\n",
      "      creating build\n",
      "      creating build\\lib.win-amd64-cpython-310\n",
      "      creating build\\lib.win-amd64-cpython-310\\msprime\n",
      "      copying msprime\\ancestry.py -> build\\lib.win-amd64-cpython-310\\msprime\n",
      "      copying msprime\\cli.py -> build\\lib.win-amd64-cpython-310\\msprime\n",
      "      copying msprime\\core.py -> build\\lib.win-amd64-cpython-310\\msprime\n",
      "      copying msprime\\demography.py -> build\\lib.win-amd64-cpython-310\\msprime\n",
      "      copying msprime\\intervals.py -> build\\lib.win-amd64-cpython-310\\msprime\n",
      "      copying msprime\\likelihood.py -> build\\lib.win-amd64-cpython-310\\msprime\n",
      "      copying msprime\\mutations.py -> build\\lib.win-amd64-cpython-310\\msprime\n",
      "      copying msprime\\pedigrees.py -> build\\lib.win-amd64-cpython-310\\msprime\n",
      "      copying msprime\\provenance.py -> build\\lib.win-amd64-cpython-310\\msprime\n",
      "      copying msprime\\species_trees.py -> build\\lib.win-amd64-cpython-310\\msprime\n",
      "      copying msprime\\_version.py -> build\\lib.win-amd64-cpython-310\\msprime\n",
      "      copying msprime\\__init__.py -> build\\lib.win-amd64-cpython-310\\msprime\n",
      "      running egg_info\n",
      "      writing msprime.egg-info\\PKG-INFO\n",
      "      writing dependency_links to msprime.egg-info\\dependency_links.txt\n",
      "      writing entry points to msprime.egg-info\\entry_points.txt\n",
      "      writing requirements to msprime.egg-info\\requires.txt\n",
      "      writing top-level names to msprime.egg-info\\top_level.txt\n",
      "      ERROR setuptools_scm._file_finders.git listing git files failed - pretending there aren't any\n",
      "      reading manifest file 'msprime.egg-info\\SOURCES.txt'\n",
      "      reading manifest template 'MANIFEST.in'\n",
      "      adding license file 'LICENSE'\n",
      "      writing manifest file 'msprime.egg-info\\SOURCES.txt'\n",
      "      copying msprime\\_msprimemodule.c -> build\\lib.win-amd64-cpython-310\\msprime\n",
      "      running build_ext\n",
      "      building 'msprime._msprime' extension\n",
      "      creating build\\temp.win-amd64-cpython-310\n",
      "      creating build\\temp.win-amd64-cpython-310\\Release\n",
      "      creating build\\temp.win-amd64-cpython-310\\Release\\git-submodules\n",
      "      creating build\\temp.win-amd64-cpython-310\\Release\\git-submodules\\tskit\n",
      "      creating build\\temp.win-amd64-cpython-310\\Release\\git-submodules\\tskit\\c\n",
      "      creating build\\temp.win-amd64-cpython-310\\Release\\git-submodules\\tskit\\c\\subprojects\n",
      "      creating build\\temp.win-amd64-cpython-310\\Release\\git-submodules\\tskit\\c\\subprojects\\kastore\n",
      "      creating build\\temp.win-amd64-cpython-310\\Release\\git-submodules\\tskit\\c\\tskit\n",
      "      creating build\\temp.win-amd64-cpython-310\\Release\\lib\n",
      "      creating build\\temp.win-amd64-cpython-310\\Release\\msprime\n",
      "      \"C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX86\\x64\\cl.exe\" /c /nologo /O2 /W3 /GL /DNDEBUG /MD -DGSL_DLL -DWIN32 -Ilwt_interface -Ilib -Igit-submodules\\tskit\\c -Igit-submodules\\tskit\\c\\tskit -Igit-submodules\\tskit\\c\\subprojects\\kastore -IC:\\Users\\Cheyanne\\AppData\\Local\\Temp\\pip-build-env-_i2v9sdy\\overlay\\Lib\\site-packages\\numpy\\core\\include -IC:\\Python310\\include -IC:\\Python310\\Include \"-IC:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.29.30133\\include\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\ucrt\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\shared\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\um\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\winrt\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\cppwinrt\" /Tcgit-submodules\\tskit\\c\\subprojects\\kastore\\kastore.c /Fobuild\\temp.win-amd64-cpython-310\\Release\\git-submodules\\tskit\\c\\subprojects\\kastore\\kastore.obj -std=c99\n",
      "      cl : Command line warning D9002 : ignoring unknown option '-std=c99'\n",
      "      kastore.c\n",
      "      git-submodules\\tskit\\c\\subprojects\\kastore\\kastore.c(28): warning C4996: 'strerror': This function or variable may be unsafe. Consider using strerror_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details.\n",
      "      git-submodules\\tskit\\c\\subprojects\\kastore\\kastore.c(588): warning C4996: 'fopen': This function or variable may be unsafe. Consider using fopen_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details.\n",
      "      \"C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX86\\x64\\cl.exe\" /c /nologo /O2 /W3 /GL /DNDEBUG /MD -DGSL_DLL -DWIN32 -Ilwt_interface -Ilib -Igit-submodules\\tskit\\c -Igit-submodules\\tskit\\c\\tskit -Igit-submodules\\tskit\\c\\subprojects\\kastore -IC:\\Users\\Cheyanne\\AppData\\Local\\Temp\\pip-build-env-_i2v9sdy\\overlay\\Lib\\site-packages\\numpy\\core\\include -IC:\\Python310\\include -IC:\\Python310\\Include \"-IC:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.29.30133\\include\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\ucrt\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\shared\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\um\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\winrt\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\cppwinrt\" /Tcgit-submodules\\tskit\\c\\tskit\\core.c /Fobuild\\temp.win-amd64-cpython-310\\Release\\git-submodules\\tskit\\c\\tskit\\core.obj -std=c99\n",
      "      cl : Command line warning D9002 : ignoring unknown option '-std=c99'\n",
      "      core.c\n",
      "      git-submodules\\tskit\\c\\tskit\\core.c(139): warning C4996: 'strerror': This function or variable may be unsafe. Consider using strerror_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details.\n",
      "      \"C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX86\\x64\\cl.exe\" /c /nologo /O2 /W3 /GL /DNDEBUG /MD -DGSL_DLL -DWIN32 -Ilwt_interface -Ilib -Igit-submodules\\tskit\\c -Igit-submodules\\tskit\\c\\tskit -Igit-submodules\\tskit\\c\\subprojects\\kastore -IC:\\Users\\Cheyanne\\AppData\\Local\\Temp\\pip-build-env-_i2v9sdy\\overlay\\Lib\\site-packages\\numpy\\core\\include -IC:\\Python310\\include -IC:\\Python310\\Include \"-IC:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.29.30133\\include\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\ucrt\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\shared\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\um\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\winrt\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\cppwinrt\" /Tcgit-submodules\\tskit\\c\\tskit\\tables.c /Fobuild\\temp.win-amd64-cpython-310\\Release\\git-submodules\\tskit\\c\\tskit\\tables.obj -std=c99\n",
      "      cl : Command line warning D9002 : ignoring unknown option '-std=c99'\n",
      "      tables.c\n",
      "      git-submodules\\tskit\\c\\tskit\\tables.c(204): warning C4996: 'strcpy': This function or variable may be unsafe. Consider using strcpy_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details.\n",
      "      git-submodules\\tskit\\c\\tskit\\tables.c(205): warning C4996: 'strcat': This function or variable may be unsafe. Consider using strcat_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details.\n",
      "      git-submodules\\tskit\\c\\tskit\\tables.c(374): warning C4996: 'strcpy': This function or variable may be unsafe. Consider using strcpy_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details.\n",
      "      git-submodules\\tskit\\c\\tskit\\tables.c(375): warning C4996: 'strcat': This function or variable may be unsafe. Consider using strcat_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details.\n",
      "      git-submodules\\tskit\\c\\tskit\\tables.c(11266): warning C4996: 'fopen': This function or variable may be unsafe. Consider using fopen_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details.\n",
      "      git-submodules\\tskit\\c\\tskit\\tables.c(11315): warning C4996: 'fopen': This function or variable may be unsafe. Consider using fopen_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details.\n",
      "      \"C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX86\\x64\\cl.exe\" /c /nologo /O2 /W3 /GL /DNDEBUG /MD -DGSL_DLL -DWIN32 -Ilwt_interface -Ilib -Igit-submodules\\tskit\\c -Igit-submodules\\tskit\\c\\tskit -Igit-submodules\\tskit\\c\\subprojects\\kastore -IC:\\Users\\Cheyanne\\AppData\\Local\\Temp\\pip-build-env-_i2v9sdy\\overlay\\Lib\\site-packages\\numpy\\core\\include -IC:\\Python310\\include -IC:\\Python310\\Include \"-IC:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.29.30133\\include\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\ucrt\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\shared\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\um\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\winrt\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\cppwinrt\" /Tcgit-submodules\\tskit\\c\\tskit\\trees.c /Fobuild\\temp.win-amd64-cpython-310\\Release\\git-submodules\\tskit\\c\\tskit\\trees.obj -std=c99\n",
      "      cl : Command line warning D9002 : ignoring unknown option '-std=c99'\n",
      "      trees.c\n",
      "      \"C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX86\\x64\\cl.exe\" /c /nologo /O2 /W3 /GL /DNDEBUG /MD -DGSL_DLL -DWIN32 -Ilwt_interface -Ilib -Igit-submodules\\tskit\\c -Igit-submodules\\tskit\\c\\tskit -Igit-submodules\\tskit\\c\\subprojects\\kastore -IC:\\Users\\Cheyanne\\AppData\\Local\\Temp\\pip-build-env-_i2v9sdy\\overlay\\Lib\\site-packages\\numpy\\core\\include -IC:\\Python310\\include -IC:\\Python310\\Include \"-IC:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.29.30133\\include\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\ucrt\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\shared\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\um\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\winrt\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\cppwinrt\" /Tclib\\avl.c /Fobuild\\temp.win-amd64-cpython-310\\Release\\lib\\avl.obj -std=c99\n",
      "      cl : Command line warning D9002 : ignoring unknown option '-std=c99'\n",
      "      avl.c\n",
      "      lib\\avl.c(223): warning C4098: 'avl_free_nodes': 'void' function returning a value\n",
      "      \"C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX86\\x64\\cl.exe\" /c /nologo /O2 /W3 /GL /DNDEBUG /MD -DGSL_DLL -DWIN32 -Ilwt_interface -Ilib -Igit-submodules\\tskit\\c -Igit-submodules\\tskit\\c\\tskit -Igit-submodules\\tskit\\c\\subprojects\\kastore -IC:\\Users\\Cheyanne\\AppData\\Local\\Temp\\pip-build-env-_i2v9sdy\\overlay\\Lib\\site-packages\\numpy\\core\\include -IC:\\Python310\\include -IC:\\Python310\\Include \"-IC:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.29.30133\\include\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\ucrt\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\shared\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\um\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\winrt\" \"-IC:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.19041.0\\cppwinrt\" /Tclib\\fenwick.c /Fobuild\\temp.win-amd64-cpython-310\\Release\\lib\\fenwick.obj -std=c99\n",
      "      cl : Command line warning D9002 : ignoring unknown option '-std=c99'\n",
      "      fenwick.c\n",
      "      lib\\fenwick.c(31): fatal error C1083: Cannot open include file: 'gsl/gsl_math.h': No such file or directory\n",
      "      error: command 'C:\\\\Program Files (x86)\\\\Microsoft Visual Studio\\\\2019\\\\BuildTools\\\\VC\\\\Tools\\\\MSVC\\\\14.29.30133\\\\bin\\\\HostX86\\\\x64\\\\cl.exe' failed with exit code 2\n",
      "      [end of output]\n",
      "  \n",
      "  note: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "  ERROR: Failed building wheel for msprime\n",
      "ERROR: Could not build wheels for msprime, which is required to install pyproject.toml-based projects\n"
     ]
    }
   ],
   "source": [
    "# installing packages to my python environment\n",
    "# NOTE: my pip is v22.0.4 but I cant seem to update to the newest 24.0\n",
    "\n",
    "#!pip install zarr\n",
    "#!pip install scipy\n",
    "#!pip install scikit-allel\n",
    "#!pip install matplotlib\n",
    "#!pip install tqdm\n",
    "#!pip install dask\n",
    "#!pip install seaborn\n",
    "#!pip install tskit\n",
    "!pip install msprime\n",
    "#!pip install pyslim\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import modules\n",
    "import numpy as np\n",
    "import zarr\n",
    "#import allel   \n",
    "#idk why but my allel doesnt seem to load?? even though ive definitely installed it as scikit-allel\n",
    "import scipy.cluster.hierarchy as sch\n",
    "import scipy.spatial\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.signal\n",
    "from scipy.ndimage import gaussian_filter1d\n",
    "from numpy.lib.stride_tricks import sliding_window_view\n",
    "from tqdm import tqdm\n",
    "import dask\n",
    "from dask.delayed import delayed\n",
    "from dask.base import compute\n",
    "from itertools import combinations\n",
    "import time\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import tskit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SEE HPC STEPS.ipynb\n",
    "\n",
    "Step 1. Generate burn in simulations establishing nucleotide diversity across a range of parameters (redo theo's work but w updated vgsc parameters)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Step 2. Conduct 27 variations of partial soft sweep simulations and output vcf files using the HPC facility. Complete x number of repeats.\n",
    "\n",
    "Size of population, N: 100, 1000, 10000\n",
    "\n",
    "Mutation rate, μ : (0.1/4N), (1/4N), (5/4N)\n",
    "\n",
    "Recombination rate, r : (0.1/4N), (1/4N), (5/4N)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "              <style>\n",
       "                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "                .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "                .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "                details {display: inline-block;}\n",
       "                summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "              </style>\n",
       "              <div class=\"tskit-table-set\">\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"padding:0;line-height:21px;\">\n",
       "                          <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "                          <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "                        </th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                      <tr><td>Trees</td><td>44241</td></tr>\n",
       "                      <tr><td>Sequence Length</td><td>70001.0</td></tr>\n",
       "                      <tr><td>Time Units</td><td>ticks</td></tr>\n",
       "                      <tr><td>Sample Nodes</td><td>20000</td></tr>\n",
       "                      <tr><td>Total Size</td><td>20.4 MiB</td></tr>\n",
       "                      <tr>\n",
       "                        <td>Metadata</td><td style=\"text-align: left;\">\n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\"></span>\n",
       "                  <details open>\n",
       "                    <summary>dict</summary>\n",
       "                    \n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\">SLiM:</span>\n",
       "                  <details >\n",
       "                    <summary>dict</summary>\n",
       "                    cycle: 50000<br/>file_version: 0.8<br/>model_type: WF<br/>name: sim<br/>nucleotide_based: False<br/>separate_sexes: False<br/>spatial_dimensionality: <br/>spatial_periodicity: <br/>stage: late<br/>tick: 50000<br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                <br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                </td></tr>\n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"line-height:21px;\">Table</th>\n",
       "                        <th>Rows</th>\n",
       "                        <th>Size</th>\n",
       "                        <th>Has Metadata</th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                    \n",
       "                  <tr>\n",
       "                    <td>Edges</td>\n",
       "                      <td>289203</td>\n",
       "                      <td>8.8 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Individuals</td>\n",
       "                      <td>10000</td>\n",
       "                      <td>978.4 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Migrations</td>\n",
       "                      <td>0</td>\n",
       "                      <td>8 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Mutations</td>\n",
       "                      <td>75042</td>\n",
       "                      <td>4.5 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Nodes</td>\n",
       "                      <td>78209</td>\n",
       "                      <td>2.8 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Populations</td>\n",
       "                      <td>1</td>\n",
       "                      <td>2.3 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Provenances</td>\n",
       "                      <td>1</td>\n",
       "                      <td>3.2 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Sites</td>\n",
       "                      <td>45943</td>\n",
       "                      <td>1.1 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "              </div>\n",
       "            </div>\n",
       "            "
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x20fa87df610>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# checking .tree outputs\n",
    "import tskit\n",
    "#tskit.load(file=\"C:/CHEYANNE STUFF/ICL Biology/Year 3/Final year project/burnin_no.19_early.trees\")\n",
    "tskit.load(file=\"C:/CHEYANNE STUFF/ICL Biology/Year 3/Final year project/burnin_no.19_early_10.trees\")\n",
    "\n",
    "\n",
    "# install pyslim !!! issue here involving msprime? pyslim cannot be imported/downloaded\n",
    "#import pyslim\n",
    "\n",
    "# read into pyslim and calculate nucleotide diversity and compare against expected nucleotide diversity (theta)\n",
    "# check that it exceeds 10% and not more than 20%\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Step 3. Read in VCF Files from soft sweep simulations, extract haplotypes using scikit_allel\n",
    "\n",
    "Working with 27 simulation variants and repeats of each variant (100?)\n",
    "\n",
    "but why only look at haplotypes of first 200 individuals? and why calculate sample frequency over specifically 400 haplotypes? is it linked to the 200 individuals = 400 genomes?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#OUT: 10000 10000 GV 20\n",
      "##fileformat=VCFv4.2\n",
      "##fileDate=20160613\n",
      "##source=SLiM\n",
      "##slimGenomePedigreeIDs=4999898,4999899,...\n",
      "##INFO=<ID=MID,Number=1,Type=Integer,Description=\"Mutation ID in SLiM\">\n",
      "##INFO=<ID=S,Number=1,Type=Float,Description=\"Selection Coefficient\">\n",
      "##INFO=<ID=DOM,Number=1,Type=Float,Description=\"Dominance\">\n",
      "##INFO=<ID=PO,Number=1,Type=Integer,Description=\"Population of Origin\">\n",
      "##INFO=<ID=TO,Number=1,Type=Integer,Description=\"Tick of Origin\">\n",
      "##INFO=<ID=MT,Number=1,Type=Integer,Description=\"Mutation Type\">\n",
      "##INFO=<ID=AC,Number=1,Type=Integer,Description=\"Allele Count\">\n",
      "##INFO=<ID=DP,Number=1,Type=Integer,Description=\"Total Depth\">\n",
      "##INFO=<ID=MULTIALLELIC,Number=0,Type=Flag,Description=\"Multiallelic\">\n",
      "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
      "#CHROM POS ID REF ALT QUAL FILTER INFO FORMAT i0 i1 i2 i3 i4 i5...\n",
      "1 1309 . A T 1000 PASS\n",
      "MID=987550;S=0;DOM=0.5;PO=2;TO=9404;MT=1;AC=11;DP=1000 GT 0|1 0|0 0|1...\n",
      "1 5995 . A T 1000 PASS\n",
      "MID=987764;S=0;DOM=0.5;PO=1;TO=9415;MT=1;AC=11;DP=1000 GT 0|1 0|0 0|1...\n",
      "...\n"
     ]
    }
   ],
   "source": [
    "# load example vcf from SLIM\n",
    "with open('example.vcf', mode='r') as vcf:\\\n",
    "\n",
    "    print(vcf.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "seed = sys.argv[1]\n",
    "#prev student: Like Hamming distance code, this was also taken from Anushka Thawani. Adaptations were made to this on the \n",
    "#high-performance computer using shell script, but this could not be represented.\n",
    "def convert(file, genome):\n",
    "    '''\n",
    "    This function extracts haplotypes sequences from a vcf file \n",
    "    Adapted from: http://alimanfoo.github.io/2018/04/09/selecting-variants.html \n",
    "    \n",
    "    Arguments:\n",
    "        file: name of vcf file (from SLiM soft sweep simulation)\n",
    "        genome: length of genome used in SLiM simulation \n",
    "        \n",
    "    Returns:\n",
    "        ht: haplotype sequences for 200 individuals\n",
    "        samp_freq: frequency of sweep mutation in sample\n",
    "        cols: used to color dendrogram\n",
    "        pos: \n",
    "\n",
    "    '''\n",
    "    \n",
    "    v = file + '.vcf'\n",
    "    z = file + '.zarr'\n",
    "    slim_sim_data = allel.read_vcf(v, fields='*')\n",
    "    allel.vcf_to_zarr(v, z, fields='*', overwrite=True)\n",
    "    data = zarr.open_group(z, mode='r')\n",
    "    \n",
    " \n",
    "    pos = allel.SortedIndex(data['variants/POS']) # Stores the ID and genomic position of each variant\n",
    "    \n",
    "    # Extract genotypes for the first 200 individuals and convert to haplotypes\n",
    "    gt = data['calldata/GT'][:,0:200] \n",
    "    ht = allel.GenotypeArray(gt).to_haplotypes()\n",
    "    \n",
    "    mutation = int((genome+1)/2) + 1  # position of sweep mutation\n",
    "    \n",
    "    \n",
    "    # Output the frequency of the sweep mutation in the sample\n",
    "    contains_sweep = pos.locate_range(mutation,mutation)\n",
    "    sweep = ht[contains_sweep]\n",
    "    sweep = np.sum(sweep, axis =0)\n",
    "    \n",
    "    samp_freq = np.sum(sweep)/400  # 400 haplotypes\n",
    "    \n",
    "    \n",
    "    # This dictionary is used later to color the dendrogram branches according to whether or not the \n",
    "    # corresponding sequence contains the sweep mutation\n",
    "    cols = {}\n",
    "    for i in range(400):\n",
    "        if sweep[i]:\n",
    "            cols[i] = \"#FF0000\" \n",
    "        else:\n",
    "            cols[i] = \"#808080\"\n",
    "    \n",
    "    return ht, pos, samp_freq, cols, sweep"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Step 4. Calculate Hij (S or homozygosity) for all pairs of haplotypes\n",
    "make a function \n",
    "homozygosity = (no. of SNPs/ length of window, L)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sliding_homozygosity(window, genome, ht, pos):\n",
    "    '''\n",
    "    This function calculates the sliding homozygosity for all pairs of haplotypes.\n",
    "    \n",
    "    Arguments:\n",
    "    window: length of sliding window\n",
    "    genome : length of genome \n",
    "    ht : vector?  of haplotype sequences from convert() function in previous codeblock\n",
    "    pos: position of variants from convert() function in previous codeblock\n",
    "\n",
    "    \n",
    "    Returns:\n",
    "    homozygosities:  homozygosity of all haplotypes in ht in a array(?)\n",
    "    '''\n",
    "    # Make empty vectors\n",
    "    homozygosities = []\n",
    "    \n",
    "    #iterate through all nucleotides in the genome\n",
    "    for x in range(0,genome):\n",
    "        #define nt ranges \n",
    "        start  = x\n",
    "        end = x + window\n",
    "        # locate the position of region around a variant Nt\n",
    "        region = pos.locate_range(start,end) \n",
    "        haplotype_region = ht [region]\n",
    "\n",
    "        #use allel.pairwise distance\n",
    "        pairwise_dist = allel.pairwise_distance(haplotype_region, metric = 'hamming' , chunked=True, blen=None) #can i chunk = true to speed up computation? \n",
    "        homozygosities = pairwise_dist/window\n",
    "\n",
    "    return homozygosities\n",
    "\n",
    "\n",
    "\n",
    "## TO DO streamlining: previous students code say to add a if condition (region =/= prev region to speed up code)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Step 5. Calculate Lij (shared haplotype length) from Hij by finding width at half maximum homozygosity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#half_max_homozygosity = (max(homozygosities) + min(homozygosities))/2\n",
    "threshold = 0.87  #setting threshold to calculate trough points in each haplotype\n",
    "\n",
    "#Defining function to find troughs, to be used in calculating Lij in another function\n",
    "def finding_troughs(smooth, pos):\n",
    "    '''\n",
    "    This function finds troughs for a pair of haplotype sequences\n",
    "    \n",
    "    Arguments:\n",
    "        sliding homozygosity: smoothed sliding window homozygosity for all pairs of sequences\n",
    "        sweeploc: position of sweep mutation in genome (same as previous function)\n",
    "        \n",
    "    Returns:\n",
    "        lower: position of breakpoint left of the sweep site\n",
    "        upper: position of breakpoint right of the sweep site\n",
    "        SHL: shared haplotype length\n",
    "    '''\n",
    "    \n",
    "    #finding troughs\n",
    "    troughs = scipy.signal.find_peaks(-smooth) #- sign inverts the graph so the 'peaks' are our troughs\n",
    "    troughs = troughs[0]     # Indexes of all troughs\n",
    "    troughs = troughs[smooth[troughs] < threshold]   # Extract troughs where homozygosity<threshold\n",
    "    \n",
    "    #finds the peaks\n",
    "    peaks = scipy.signal.find_peaks(smooth)\n",
    "    peaks = peaks[0]  #indexes peaks\n",
    "    \n",
    "    # Find positions of troughs flanking sweep site\n",
    "    bp = np.searchsorted(troughs,pos)  #search sorted finds index of position where mutation should be inserted in order to maintain the same order\n",
    "    lower = troughs[bp - 1] #index of sweep site -1\n",
    "    upper = troughs[bp]  #index of sweep site\n",
    "    \n",
    "    # Find the average peak position around the sweep site\n",
    "    highest = peaks[(peaks >= lower) & (peaks <= upper)]\n",
    "    if highest.size != 0:\n",
    "        highest = np.mean(highest)\n",
    "    else: \n",
    "        highest = (lower+upper)/2\n",
    "    \n",
    "    \n",
    "    lower = (lower+highest)/2\n",
    "    upper = (upper+highest)/2\n",
    "\n",
    "    SHL = upper - lower\n",
    "    \n",
    "    return int(lower), int(upper), SHL\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## function using each haplotype pair to return SHL and find lower and upper limits of SHL\n",
    "# uses finding_troughs()\n",
    "def find_breakpoint(haplotype_pair):\n",
    "    '''\n",
    "    For a pair of sequences, this function smoothes the sliding homozygosity and returns the SHL\n",
    "    Arguments:haplotype_pair\n",
    "        haplotype_pair: a pair of haplotype sequences\n",
    "        \n",
    "    Returns:\n",
    "        lower: position of breakpoint left of the sweep site\n",
    "        upper: position of breakpoint right of the sweep site\n",
    "        SHL: shared haplotype length\n",
    "    '''\n",
    "    \n",
    "    mutation_pos = mutation \n",
    "    smooth = gaussian_filter1d(haplotype_pair, points_g)\n",
    "    try:\n",
    "        lower, upper, SHL = finding_troughs(smooth, mutation_pos)\n",
    "    except IndexError:\n",
    "        lower = -1.3\n",
    "        upper = -1.3\n",
    "        SHL = -1.3\n",
    "        \n",
    "    return lower, upper, SHL"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Step 6. Calculate Tij (Time to common ancestor) from Lij and Kij (which is no. of SNPs) on each shared length (haplotype)\n",
    "\n",
    "𝜏_𝑖𝑗=(𝑘_𝑖𝑗+1)/(2ℓ_𝑖𝑗 (𝑟+𝜇))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculating Kij (No. of SNPs in each SHL)\n",
    "\n",
    "def calculating_kij(n,gts,ht,result_find_breakpoint,pos):\n",
    "    '''\n",
    "    This function finds the number of SNPs over the shared haplotype length for all pairs of haplotype sequences\n",
    "    \n",
    "    Arguments:\n",
    "        n: number of haplotype sequences\n",
    "        gts: number of haplotype pairs\n",
    "        ht: haplotype sequnces\n",
    "        result_find_breakpoint: output from find_breakpoint function\n",
    "        \n",
    "    Returns:\n",
    "        diffs: number of SNPs for all pairs of haplotype sequences\n",
    "        \n",
    "    '''\n",
    "    pairwise = []\n",
    "    for combo in combinations(list(range(0,n)), 2): \n",
    "        pairwise.append(combo)\n",
    "\n",
    "    diffs = np.empty(shape=(gts),dtype=np.float32)\n",
    "    for i in range(gts):\n",
    "        pair = ht[:,pairwise[i]]\n",
    "        try:\n",
    "            start = result_find_breakpoint[i,1]\n",
    "            stop = result_find_breakpoint[i,2]\n",
    "\n",
    "            window_pos = pos.locate_range(start, stop)\n",
    "            window = pair[window_pos]\n",
    "\n",
    "            d = allel.pairwise_distance(window, metric = \"hamming\")\n",
    "\n",
    "            diffs[i]=d \n",
    "\n",
    "        except KeyError:\n",
    "            diffs[i]=-1.3 \n",
    "    \n",
    "    return diffs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Calculating Tij, Time to Common Ancestor (TMRCA) using mutation rate and number of SNPs\n",
    "# 𝜏_𝑖𝑗=(𝑘_𝑖𝑗+1)/(2ℓ_𝑖𝑗 (𝑟+𝜇))\n",
    "\n",
    "def calculating_Tij(file, genome, pop_size, diffs, mu, recombination_rate):\n",
    "    '''\n",
    "    This function calculates Tij, Time to Common Ancestor (TMRCA) using mutation rate and number of SNPs for each haplotype pair.\n",
    "    It uses the convert() to convert files from vcf \n",
    "    It uses the sliding_heterozygosity() to calculate heterozygosity in the window for all pairs of haplotypes\n",
    "\n",
    "    \n",
    "    Arguments:\n",
    "        file: name of vcf file\n",
    "        genome_length: length of genome (in SLiM simulation)             ##can loop?\n",
    "        pop: effective population size (in SLiM simulation)       ##can loop?\n",
    "        diffs: Kij , the number of SNPs in a given genome\n",
    "        mu: mutation rate\n",
    "        recombination_rate: recombination rate \n",
    "        ht : haplotypes (what variable structure?)\n",
    "\n",
    "        window: length of sliding window\n",
    "        threshold: threshold above which troughs are ignored\n",
    "        points: number of points to use for 1D-gaussian filter (see scipy documentation)\n",
    "        \n",
    "    Returns:\n",
    "        Tij: Time to Common Ancestor (TMRCA)\n",
    "  \n",
    "    '''\n",
    "    \n",
    "    #print(file)\n",
    "\n",
    "    global mutation\n",
    "    mutation = int((genome+1)/2) \n",
    "    global threshold\n",
    "    global points_g\n",
    "    points_g = points\n",
    "    \n",
    "    # Extract haplotype sequences from .vcf file\n",
    "    ht, pos, samp_freq, cols, sweep = convert(file, genome)\n",
    "\n",
    "    \n",
    "    # Calculate sliding homozygosity for all pairs of haplotype sequences\n",
    "    n = 400 #number of haplotypes \n",
    "    gts = int((n*(n-1))/2)\n",
    "    homozygosities = sliding_homozygosity(window, mutation, genome, ht, pos, gts)\n",
    "\n",
    "    \n",
    "    # Find SHL for all pairs of haplotype sequences \n",
    "    hom_dask = dask.array.from_array(homozygosities, chunks=(genome,1)) # creates a dask array\n",
    "    homozygosities = []\n",
    "    results = dask.array.apply_along_axis(find_breakpoint, 0, hom_dask) #applies find_breakpoint() along the array\n",
    "    results_computed = results.compute()\n",
    "\n",
    "    # Manipulating the dataframe to make it easier to process\n",
    "    results_computed = np.transpose(results_computed)\n",
    "    index = np.asarray(range(0,gts))\n",
    "    index = np.expand_dims(index, axis=0)\n",
    "    results_computed_1 = np.concatenate((index.T, results_computed), axis=1)\n",
    "    \n",
    "    \n",
    "    # Calculate the TMRCA from the SHLs and number of SNPs\n",
    "    recombination_rate = recombination_rate/(2*pop)\n",
    "    mu = mu/(100*pop)\n",
    "    shls = results_computed_1[:,3]   # SHLs for all pairs of haplotype sequences \n",
    "    shls[shls<=0] = genome\n",
    "    diffs = find_snp(n, gts, ht, results_computed_1, pos)  # SNPs for all pairs of haplotype sequences \n",
    "    Tij = (1+(diffs*shls))/(2*shls*(recombination_rate + mu)) # TMRCA metric for all pairs of haplotype sequences \n",
    "\n",
    "    \n",
    "    # Remove negative and non-integer TMRCA values\n",
    "    impute = np.nanmean(Tij)        #impute is the mean of SNP array without any NAN values\n",
    "    x = np.isfinite(Tij)            #x is a boolean mask array of only finite values (cannot be infinite or NAN)\n",
    "    for i in np.where(x == 0)[0]:   #for all indices where there is a non-finite number:\n",
    "        Tij[i] = impute             #replace NaN with inpute value\n",
    "    Tij[Tij<=0] = impute            # replace all negative numbers wih inpute\n",
    "\n",
    "    return Tij, cols\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Step 7. Plotting TMRCA dendrogram \n",
    "- prev student ran into problems here with generating the dendrogram coloured tips\n",
    "- Untested to see if I have resolved previous student problem\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def analysis(file,genome,pop,window,threshold,points,r=1,u=1): \n",
    "    '''\n",
    "    This function plots a dendrogram and colours red the tips that have the sweep mutations. \n",
    "    Only the output vcf from the SLIM simulation is input.\n",
    "\n",
    "    Arguments:\n",
    "    file: output vcf file from SLIM simulation\n",
    "    genome: genome length of simulation\n",
    "    pop: effective population size (in SLiM simulation)\n",
    "        window: length of sliding window\n",
    "        threshold: threshold above which troughs are ignored\n",
    "        points: number of points to use for 1D-gaussian filter (see scipy documentation)\n",
    "        r: recombination rate\n",
    "        u: mutation rate \n",
    "    cols: colours haplotype branches red if they have sweep mutation. From convert().\n",
    "\n",
    "\n",
    "    Returns:\n",
    "    output dendrogram in pdf\n",
    "    '''\n",
    "    #all convert(), etc etc to end up with tij\n",
    "    Tij, cols = calculating_Tij(file, genome, pop_size, diffs, mu, recombination_rate)\n",
    "    \n",
    "    # Hierachical Clustering, store in Z\n",
    "    Z = sch.linkage(Tij, method = 'average') #why do we use the Farthest Point Algorithm? changed to average (UPGMA) , check after\n",
    "    \n",
    "\n",
    "    ## Plot dendrogram without colouring branches\n",
    "    # updating matplotlib font settings\n",
    "    matplotlib.rcParams.update({'font.size': 24})\n",
    "    fig = plt.figure(figsize=(30, 12))\n",
    "    gs = matplotlib.gridspec.GridSpec(2, 1, hspace=0.1, wspace=1, height_ratios=(1,1))\n",
    "\n",
    "    ax_dend = fig.add_subplot(gs[0, 0])\n",
    "    sns.despine(ax=ax_dend, offset=5, bottom=True, top=True)\n",
    "    dd = sch.dendrogram(Z,color_threshold=0,above_threshold_color='#808080',ax=ax_dend) # if above colour threshold, set colour to grey \n",
    "\n",
    "    ls = []\n",
    "    for leaf, leaf_color in zip(plt.gca().get_xticklabels(), dd[\"leaves_color_list\"]):   #leaves_color_list is A list of color names. The k’th element represents the color of the k’th leaf.\n",
    "        leaf.set_color(cols[int(leaf.get_text())])\n",
    "        ls.append(int(leaf.get_text()))\n",
    "\n",
    "    ax_dend.set_ylabel('Haplotype age/generations',fontsize=24)\n",
    "    ax_dend.set_title('Haplotype clusters',fontsize=24)\n",
    "\n",
    "\n",
    "\n",
    "    # Plot dendrogram and colour branches\n",
    "    ax_dend_2 = fig.add_subplot(gs[1, 0])\n",
    "    \n",
    "    dflt_col = \"#808080\"\n",
    "    \n",
    "    link_cols = {}\n",
    "    for i, i12 in enumerate(Z[:,:2].astype(int)):\n",
    "        c1, c2 = (link_cols[x] if x > len(Z) else cols[x] for x in i12)\n",
    "        link_cols[i+1+len(Z)] = c1 if c1 == c2 else dflt_col\n",
    "\n",
    "    sns.despine(ax=ax_dend_2, offset=5, bottom=True, top=True)\n",
    "    dd = sch.dendrogram(Z,link_color_func=lambda x: link_cols[x], ax=ax_dend_2)\n",
    "\n",
    "    ls = []\n",
    "    for leaf, leaf_color in zip(plt.gca().get_xticklabels(), dd[\"leaves_color_list\"]):\n",
    "        leaf.set_color(cols[int(leaf.get_text())])\n",
    "        ls.append(int(leaf.get_text()))\n",
    "\n",
    "    ax_dend_2.set_ylabel('Haplotype age/generations',fontsize=24)\n",
    "    \n",
    "    \n",
    "    # Save dendrogram\n",
    "    output = 'accurate_' + file + '.pdf'\n",
    "    plt.savefig(output)  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Step 4. Estimate n hat by counting number of origins in dendrogram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# count number of origins (haplotypes) for all files and save in a list\n",
    "\n",
    "def sum_origins (number_of_files, files):\n",
    "    '''\n",
    "    takes the number of simulations, the .vcf files and \n",
    "    calculates the number of origins per simulation and saves it in a list.\n",
    "\n",
    "\n",
    "    returns:\n",
    "\n",
    "    '''\n",
    "    # search if can use sch.dendrogram to get number of haplotype origins \n",
    "    origins = []\n",
    "    for i in range (number_of_files):\n",
    "        read.file ( file + i + .TREES) as file: \n",
    "\n",
    "            #if _________ =___________:\n",
    "                #calculate  number of origins as file_origins\n",
    "        #attach file_origins to origins\n",
    "    return print (origins)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Step 5. Calculate Ne hat by maximum likelihood estimation\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot all estimated numbers of origins\n",
    "\n",
    "# find a function that plots and finds maximum likelihood and the 95% confidence intervals\n",
    "# calculating confidence interval by \n",
    "# 1. log ne hat and find maximum which should be the maximum of the graph as well\n",
    "# average(origins) ?\n",
    "\n",
    "\n",
    "def NeHat(csv_file_path, array_num):\n",
    "    '''\n",
    "    This function plots the number of independent gene loci/origins (L) against expected to check\n",
    "    \n",
    "    Arguments:\n",
    "        csv_file_path: filepath of the csv used to manually record down the number of origins per simulation\n",
    "        array_num : combination number of this simulation (1-27)\n",
    "        \n",
    "    Returns:\n",
    "        confidence_intervals: 95% confidence intervals of NeHat being correct\n",
    "        NeHat: NeHat (at ln(Lmax))\n",
    "\n",
    "    ''' \n",
    "    #load csv\n",
    "    origins = pd.read_csv('C:/CHEYANNE STUFF/ICL Biology/Year 3/Final year project/Estimating-effective-population-size-from-resistance-loci/simulations/origins.csv')\n",
    "    index = array_num +1\n",
    "\n",
    "    # log the heterozygosity (ln(L))\n",
    "    # At max ln(L), is Ne Hat\n",
    "    Lmax = df['Heterozygosity'].max()\n",
    "    df[\"ln(L)\"] = np.log(df['Heterozygosity'])\n",
    "    ln_Lmax = np.log(Lmax) \n",
    "    \n",
    "    # at [ln(Lmax)-2], is Nelow and NeHigh the confidence intervals\n",
    "    confidence_cutoff = ln_Lmax -2 \n",
    "    #NeLow =       ##how to find best fit line and intersection?\n",
    "\n",
    "\n",
    "\n",
    "    # plot to look at confidence intervals, NeHat\n",
    "    Lmax = max (df['Heterozygosity'])\n",
    "    ln_Lmax = np.log(Lmax) \n",
    "    plt.plot('cycle','Heterozygosity', data = df)\n",
    "    plt.axhline(y =ln_Lmax, color='r', linestyle='-') #ln_Lmax\n",
    "    plt.axhline(y=confidence_cutoff, color='r', linestyle='-') #confidence interval cutoff\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "\n",
    "    return NeLow, NeHigh, NeHat"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
